core/modules/wpublish/QueryChunkStatistics.cc

```diff
@@ -210,6 +211,53 @@ std::ostream& operator<<(std::ostream& os, QueryStatistics const& q) {
 }
 
 
+/// Remove all Tasks belonging to the user query 'qId' from the queue of 'sched'.
+/// If 'sched' is null, then Tasks will be removed from the queue of whatever scheduler
+/// they are queued on.
+/// Tasks that are already running continue, but are marked as complete on their
+/// current scheduler. Stopping and rescheduling them would be difficult at best, and
+/// probably not very helpful.
+std::vector<wbase::Task::Ptr>
+QueryChunkStatistics::removeQueryFrom(QueryId const& qId,
+                                      std::shared_ptr<wsched::SchedulerBase> const& sched) {
+
+    std::vector<wbase::Task::Ptr> removedList; // Return value;
+
+    // Find the user query.
+    std::unique_lock<std::mutex> lock(_queryStatsMtx);
+    auto query = _queryStats.find(qId);
+    if (query == _queryStats.end()) {
+        LOGS(_log, LOG_LVL_DEBUG, QueryIdHelper::makeIdStr(qId) << " was not found by removeQueryFrom");
+        return removedList;;
+    }
+    lock.unlock();
+
+    // Remove Tasks from their scheduler put them on 'removedList', but only if their Scheduler is the same
+    // as 'sched'or if sched == nullptr.
+    auto &taskMap = query->second->_taskMap;
+    std::vector<wbase::Task::Ptr> taskList;
+    {
+        std::lock_guard<std::mutex> taskLock(query->second->_qStatsMtx);
+        for (auto const& elem : taskMap) {
+            taskList.push_back(elem.second);
+        }
+    }
+
+    for(auto const& task : taskList) {
+        auto baseTaskSched = std::dynamic_pointer_cast<wsched::SchedulerBase>(task->getTaskScheduler());
```

Comment:
I added a nullptr check.


On 07/27/16 14:49, Andy wrote:
>
> In core/modules/wpublish/QueryChunkStatistics.cc 
> <https://github.com/lsst/qserv/pull/243#discussion_r72530718>:
>
> > +    }
> > +    lock.unlock();
> > +
> > +    // Remove Tasks from their scheduler put them on 'removedList', but only if their Scheduler is the same
> > +    // as 'sched'or if sched == nullptr.
> > +    auto &taskMap = query->second->_taskMap;
> > +    std::vector<wbase::Task::Ptr> taskList;
> > +    {
> > +        std::lock_guard<std::mutex> taskLock(query->second->_qStatsMtx);
> > +        for (auto const& elem : taskMap) {
> > +            taskList.push_back(elem.second);
> > +        }
> > +    }
> > +
> > +    for(auto const& task : taskList) {
> > +        auto baseTaskSched = std::dynamic_pointer_cast<wsched::SchedulerBase>(task->getTaskScheduler());
>
> If |baseTaskSched| can be nullptr then code below will crash if 
> |sched| is nullptr too.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/lsst/qserv/pull/243/files/af3ba79d72fa0ea4d85aaf89b13e9c6907dfdf74#r72530718>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGQRafTrWhYy_Yl8Ja1Mhm1WC9_jJuTdks5qZ9JlgaJpZM4JUiuv>.
>